presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptorBuilder.java

```diff
@@ -0,0 +1,154 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static com.facebook.presto.orc.metadata.ColumnEncoding.DEFAULT_SEQUENCE_ID;
+import static com.facebook.presto.orc.metadata.OrcType.OrcTypeKind.STRUCT;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public final class StreamDescriptorBuilder
+{
+    private static final int ROOT_ID = 0;
+
+    private StreamDescriptorBuilder()
+    {
+    }
+
+    public static StreamDescriptor createStreamDescriptor(List<OrcType> types, OrcDataSource dataSource)
+    {
+        ImmutableMap.Builder<Integer, StreamProperty> propertiesBuilder = ImmutableMap.builderWithExpectedSize(types.size());
+        addOrcType("", "", ROOT_ID, types, propertiesBuilder);
+        AllStreams allStreams = new AllStreams(dataSource, propertiesBuilder.build());
+        return new StreamDescriptor(ROOT_ID, DEFAULT_SEQUENCE_ID, allStreams);
+    }
+
+    private static void addOrcType(String parentStreamName, String fieldName, int typeId, List<OrcType> types, ImmutableMap.Builder<Integer, StreamProperty> propertiesBuilder)
+    {
+        OrcType type = types.get(typeId);
+
+        if (!fieldName.isEmpty()) {
+            parentStreamName += "." + fieldName;
+        }
+
+        ImmutableList.Builder<Integer> nestedStreamIdsBuilder = ImmutableList.builderWithExpectedSize(type.getFieldCount());
+        if (type.getOrcTypeKind() == STRUCT) {
+            for (int i = 0; i < type.getFieldCount(); ++i) {
+                nestedStreamIdsBuilder.add(type.getFieldTypeIndex(i));
+                addOrcType(parentStreamName, type.getFieldName(i), type.getFieldTypeIndex(i), types, propertiesBuilder);
+            }
+        }
+        else if (type.getOrcTypeKind() == OrcType.OrcTypeKind.LIST) {
+            nestedStreamIdsBuilder.add(type.getFieldTypeIndex(0));
+            addOrcType(parentStreamName, "item", type.getFieldTypeIndex(0), types, propertiesBuilder);
+        }
+        else if (type.getOrcTypeKind() == OrcType.OrcTypeKind.MAP) {
+            nestedStreamIdsBuilder.add(type.getFieldTypeIndex(0));
+            nestedStreamIdsBuilder.add(type.getFieldTypeIndex(1));
+            addOrcType(parentStreamName, "key", type.getFieldTypeIndex(0), types, propertiesBuilder);
+            addOrcType(parentStreamName, "value", type.getFieldTypeIndex(1), types, propertiesBuilder);
+        }
+        StreamProperty streamProperty = new StreamProperty(parentStreamName, type, fieldName, nestedStreamIdsBuilder.build());
+        propertiesBuilder.put(typeId, streamProperty);
+    }
+
+    public static class StreamProperty
+    {
+        private final String streamName;
+        private final OrcType orcType;
+        private final String fieldName;
+        private final List<Integer> nestedStreamIds;
```

Comment:
If you are looking to save even a bit more memory, you can look into IntArrayList/IntImmutableList from fastutil. But probably it's not worth it in this case.

Reply:
using raw byte arrays, for empty arrays we have to pool. ImmutableList.of() uses the same static array and saves memory. So will use the ImmutableList for now.